source/iterator/snapshot.go

```diff
@@ -0,0 +1,176 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-hubspot/hubspot"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+)
+
+// Snapshot is an implementation of a Snapshot iterator for the HubSpot API.
+type Snapshot struct {
+	hubspotClient *hubspot.Client
+	resource      string
+	bufferSize    int
+	pollingPeriod time.Duration
+	records       chan sdk.Record
+	errC          chan error
+	position      *Position
+}
+
+// SnapshotParams is an incoming params for the [NewSnapshot] function.
+type SnapshotParams struct {
+	HubSpotClient *hubspot.Client
+	Resource      string
+	BufferSize    int
+	PollingPeriod time.Duration
+	Position      *Position
+}
+
+// NewSnapshot creates a new instance of the [Snapshot].
+func NewSnapshot(ctx context.Context, params SnapshotParams) *Snapshot {
+	snapshot := &Snapshot{
+		hubspotClient: params.HubSpotClient,
+		resource:      params.Resource,
+		bufferSize:    params.BufferSize,
+		pollingPeriod: params.PollingPeriod,
+		records:       make(chan sdk.Record, params.BufferSize),
+		errC:          make(chan error, 1),
+		position:      params.Position,
+	}
+
+	go snapshot.poll(ctx)
+
+	return snapshot
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+func (s *Snapshot) HasNext(ctx context.Context) (bool, error) {
+	return len(s.records) > 0, nil
+}
+
+// Next returns the next record.
+func (s *Snapshot) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, fmt.Errorf("context cancelled: %w", ctx.Err())
+
+	case err := <-s.errC:
+		return sdk.Record{}, fmt.Errorf("async error: %w", err)
+
+	case record := <-s.records:
+		return record, nil
+	}
+}
+
+// poll polls items at the specified time intervals.
+func (s *Snapshot) poll(ctx context.Context) {
```

Comment:
As I understand we get resources from hubspot each 5 second using filter: all records after last  processed id, and push it one by one to record channel.  Is it possible to get duplicates  on this case if polling period will be short? for example  we get 20 records with ids (1,2,3,4 .....19,20) . We publish to record chan 15 records and ticker start new iteration and we will get records with ids (15, 16,.......35)

Reply:
Hm, each time we send an item to the `records` channel we update the iterator's position. And when we start polling new items we use the iterator's position which contains an id of the last processed element. So I think duplicates are not possible in this case. I'll check it.

Reply:
thank you!

Reply:
Quick recap: the `time.Ticker` will wait until the `loadRecords` function is finished. So we won't call the `loadRecords` multiple times at the same time.